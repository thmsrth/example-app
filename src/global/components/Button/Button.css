:root {
  /** Colors **/
  --button-background-color: var(--color-extra-light-grey);
  --button-color: var(--color-black);
  --button-border-color: var(--color-light-grey);

  --button-hover-background-color: var(--color-white);
  --button-hover-border-color: var(--color-light-grey);
  --button-hover-color: var(--color-black);

  --button-disabled-background-color: var(--state-disabled-background-color);
  --button-disabled-color: var(--state-disabled-color);

  /** Layout **/
  --button-border-radius: var(--border-radius);
  --button-space: var(--space);
  --button-padding: calc(var(--button-space) / 3) calc(var(--button-space) / 2);

  /** Typography **/
  --button-font-family: inherit;
  --button-font-weight: normal;
  --button-font-size: 1.4rem;
  --button-line-height: default;
}

.Button {
  opacity: 1;
  background-color: var(--button-background-color);
  color: var(--button-color);
  border: 1px solid var(--button-border-color);
  border-radius: var(--button-border-radius);

  display: flex;
  align-items: center;
  justify-content: center;

  margin: 0;
  padding: var(--button-padding);

  font-family: var(--button-font-family);
  font-weight: var(--button-font-weight);
  font-size: var(--button-font-size);
  line-height: var(--button-line-height);
  white-space: nowrap;
  text-transform: none;

  background-image: none;
  user-select: none;
  touch-action: manipulation;
  -webkit-appearance: button;
  overflow: visible;
}

.Button:hover {
  cursor: pointer;
  opacity: 0.8;
}

.Button:active {
  box-shadow: inset 3px 3px 5px var(--button-border-color);
  background-color: var(--button-hover-background-color);
  border-color: var(--button-hover-border-color);
  color: var(--button-hover-color);
}

.Button--disabled {
  background-color: var(--button-disabled-background-color);
}

.Button--disabled .Button__icon,
.Button--disabled .Button__text {
  color: var(--button-disabled-color);
}

.Button--disabled:hover {
  cursor: not-allowed;
  opacity: 1;
}

.Button__icon,
.Button__text {
  display: inline-block;
}

.Button__icon {
  margin-right: calc(var(--button-space) / 3);
}
