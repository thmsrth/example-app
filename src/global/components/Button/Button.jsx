import React from 'react';
import PropTypes from 'prop-types';
import Icon from '../Icon';

import { bemCls } from '../../utils/ClassNameHelpers';

import './Button.css';

const Button = ({
  className, text, type, iconName, disabled, onClick, loading, tooltip,
}) => {
  if ((onClick || type === 'submit') && (text || iconName)) {
    return (
      // eslint-disable-next-line react/button-has-type
      <button
        type={type}
        className={`Button ${className} ${
          bemCls(`Button ${className}`, '--disabled', disabled)}`}
        onClick={onClick}
        disabled={disabled || loading}
        title={tooltip}
      >
        {iconName
          && (
          <div className={`${bemCls(`Button ${className}`, '__icon')} ${
            bemCls(`Button ${className}`, '__icon--disabled', disabled)}`}
          >
            {loading ? <Icon name="spinner" className="fa-spin" /> : <Icon name={iconName} />}
          </div>
          )
        }
        {text
          && (
          <div className={`${bemCls(`Button ${className}`, '__text')} ${
            bemCls(`Button ${className}`, '__text--disabled', disabled)}`}
          >
            {text}
          </div>
          )
        }
      </button>
    );
  }
  return null;
};

export default Button;

Button.propTypes = {
  className: PropTypes.string,
  text: PropTypes.string,
  tooltip: PropTypes.string,
  type: PropTypes.string,
  iconName: PropTypes.string,
  disabled: PropTypes.bool,
  loading: PropTypes.bool,
  onClick: PropTypes.func,
};

Button.defaultProps = {
  className: '',
  text: '',
  tooltip: '',
  type: 'button',
  iconName: '',
  disabled: false,
  loading: false,
  onClick: undefined,
};
