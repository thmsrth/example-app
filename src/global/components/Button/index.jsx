import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Button from './Button';

class ButtonWrapper extends Component {
  constructor() {
    super();
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick() {
    const { onClick, params } = this.props;
    onClick(params);
  }

  render() {
    return (
      <Button
        {...this.props}
        onClick={this.handleClick}
      />
    );
  }
}

ButtonWrapper.propTypes = {
  onClick: PropTypes.func,
  params: PropTypes.objectOf(PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ])),
};

ButtonWrapper.defaultProps = {
  onClick: () => {},
  params: {},
};

export default ButtonWrapper;
