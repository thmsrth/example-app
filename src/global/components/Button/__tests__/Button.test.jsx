import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import Button from '../Button';

const props = {
  className: 'myClassname',
  onClick: () => {},
};
const text = 'Mytext';
const iconName = 'chevron-up';

describe('Button', () => {
  it('should render the component correctly with text and icon', () => {
    const tree = shallow(<Button {...props} text={text} iconName={iconName} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component without text if iconName is defined and not blank', () => {
    const tree = shallow(<Button {...props} iconName={iconName} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component without icon if text is defined and not blank', () => {
    const tree = shallow(<Button {...props} text={text} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component for disabled state', () => {
    const tree = shallow(<Button {...props} text={text} iconName={iconName} disabled />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should not render the component if text and iconName are undefined', () => {
    const tree = shallow(<Button {...props} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should not render the component if text and iconName are blank', () => {
    const tree = shallow(<Button {...props} text="" iconName="" />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component if onClick is undefined and type is submit', () => {
    const tree = shallow(<Button {...props} text={text} onClick={undefined} type="submit" />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should not render the component if onClick is undefined and type not submit', () => {
    const tree = shallow(<Button {...props} text={text} onClick={undefined} type="button" />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });

  it('should render the component for loading state', () => {
    const tree = shallow(<Button {...props} text={text} iconName={iconName} loading />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
