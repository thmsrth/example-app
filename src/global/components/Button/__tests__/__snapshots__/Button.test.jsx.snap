// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should not render the component if onClick is undefined and type not submit 1`] = `""`;

exports[`Button should not render the component if text and iconName are blank 1`] = `""`;

exports[`Button should not render the component if text and iconName are undefined 1`] = `""`;

exports[`Button should render the component correctly with text and icon 1`] = `
<button
  className="Button myClassname "
  disabled={false}
  onClick={[Function]}
  title=""
  type="button"
>
  <div
    className="Button__icon myClassname__icon "
  >
    <Icon
      className=""
      name="chevron-up"
    />
  </div>
  <div
    className="Button__text myClassname__text "
  >
    Mytext
  </div>
</button>
`;

exports[`Button should render the component for disabled state 1`] = `
<button
  className="Button myClassname Button--disabled myClassname--disabled"
  disabled={true}
  onClick={[Function]}
  title=""
  type="button"
>
  <div
    className="Button__icon myClassname__icon Button__icon--disabled myClassname__icon--disabled"
  >
    <Icon
      className=""
      name="chevron-up"
    />
  </div>
  <div
    className="Button__text myClassname__text Button__text--disabled myClassname__text--disabled"
  >
    Mytext
  </div>
</button>
`;

exports[`Button should render the component for loading state 1`] = `
<button
  className="Button myClassname "
  disabled={true}
  onClick={[Function]}
  title=""
  type="button"
>
  <div
    className="Button__icon myClassname__icon "
  >
    <Icon
      className="fa-spin"
      name="spinner"
    />
  </div>
  <div
    className="Button__text myClassname__text "
  >
    Mytext
  </div>
</button>
`;

exports[`Button should render the component if onClick is undefined and type is submit 1`] = `
<button
  className="Button myClassname "
  disabled={false}
  title=""
  type="submit"
>
  <div
    className="Button__text myClassname__text "
  >
    Mytext
  </div>
</button>
`;

exports[`Button should render the component without icon if text is defined and not blank 1`] = `
<button
  className="Button myClassname "
  disabled={false}
  onClick={[Function]}
  title=""
  type="button"
>
  <div
    className="Button__text myClassname__text "
  >
    Mytext
  </div>
</button>
`;

exports[`Button should render the component without text if iconName is defined and not blank 1`] = `
<button
  className="Button myClassname "
  disabled={false}
  onClick={[Function]}
  title=""
  type="button"
>
  <div
    className="Button__icon myClassname__icon "
  >
    <Icon
      className=""
      name="chevron-up"
    />
  </div>
</button>
`;
